api/src/services/security-service.ts

```diff
@@ -330,53 +330,94 @@ export class SecurityService extends DBService {
+   * @param {number[]} removeRuleIds IDs of the rules which will be removed after the patch operation
+   * @return {*}  {Promise<void>}
    * @memberof SecurityService
    */
-  async applySecurityRulesToSubmissionFeatures(
-    features: number[],
-    rules: number[],
-    override = false
-  ): Promise<SubmissionFeatureSecurityRecord[]> {
-    if (override) {
-      // we want to override any security rules present and can achieve this by remove everything first
-      await this.securityRepository.removeSecurityRulesFromSubmissionFeatures(features);
+  async patchSecurityRulesOnSubmissionFeatures(
+    submissionFeatureIds: number[],
+    applyRuleIds: number[],
+    removeRuleIds: number[]
+  ): Promise<void> {
+    defaultLog.debug({ label: 'patchSecurityRulesOnSubmissionFeatures', applyRuleIds, removeRuleIds });
+
+    if (!submissionFeatureIds.length) {
+      return;
     }
 
-    return this.securityRepository.applySecurityRulesToSubmissionFeatures(features, rules);
+    if (removeRuleIds.length > 0) {
+      await this.securityRepository.removeSecurityRulesFromSubmissionFeatures(submissionFeatureIds, removeRuleIds);
+    }
+
+    if (applyRuleIds.length > 0) {
+      await this.securityRepository.applySecurityRulesToSubmissionFeatures(submissionFeatureIds, applyRuleIds);
+    }
   }
 
   /**
-   * Removes all security rules for the given set of submission feature ids
+   * Removes the given security rules from the given set of submission feature ids. If
+   * no security rules ID is provided, all security rules will be removed for the given set
+   * of subission features.
+   *
+   * @param {number[]} submissionFeatureIds
+   * @param {number[]} [removeRuleIds]
    *
    * @return {*}  {Promise<SubmissionFeatureSecurityRecord[]>}
    * @memberof SecurityService
    */
-  async removeSecurityRulesFromSubmissionFeatures(features: number[]): Promise<SubmissionFeatureSecurityRecord[]> {
-    return this.securityRepository.removeSecurityRulesFromSubmissionFeatures(features);
+  async removeSecurityRulesFromSubmissionFeatures(
+    submissionFeatureIds: number[],
+    removeRuleIds?: number[]
+  ): Promise<SubmissionFeatureSecurityRecord[]> {
+    if (!submissionFeatureIds.length) {
+      return [];
+    }
+
+    if (!removeRuleIds) {
+      return this.securityRepository.removeAllSecurityRulesFromSubmissionFeatures(submissionFeatureIds);
+    }
+
+    return this.securityRepository.removeSecurityRulesFromSubmissionFeatures(submissionFeatureIds, removeRuleIds);
   }
 
   /**
    * Gets Submission Feature Security Records for a given set of submission feature ids
    *
-   * @param {number[]} features
+   * @param {number[]} submissionFeatureIds
+   * @return {*}  {Promise<SecurityRuleRecord[]>}
+   * @memberof SecurityService
+   */
+  async getSecurityRulesForSubmissionFeatures(
+    submissionFeatureIds: number[]
+  ): Promise<SubmissionFeatureSecurityRecord[]> {
+    if (!submissionFeatureIds.length) {
```

Comment:
Same questions as above. Is this check required when the GET request API doc forces a minimum 1 id.

Reply:
Is this function even used in the current state?

Reply:
This function is not currently being used, because a new function is providing similar enough functionality. But because this function is adjacent to a problem that is still evolving in scope (applying and removing security) I think it makes sense to keep it around (at least for now).